test/es7/asyncawait-functionality.js

```diff
@@ -1037,21 +1037,46 @@ var tests = [
                 }
             );
         }
+    },
+    {
+        name: "`then` is called with both onFulfilled and onRejected",
+        body: function (index) {
+            async function bar() {
+                throw new Error("Whoops");
+            }
+
+            async function foo() {
+                try {
+                    await bar();
+                } catch (e){
+                    echo(`Test #${index} - Success caught the expected exception`);
+                }
+            }
+
+            var oldThen = Promise.prototype.then;
+            Promise.prototype.then = function(thenx, catchx) {
+                echo(`Test #${index} - then: ${!!thenx}, catch: ${!!catchx}`)
+                return oldThen.apply(this, arguments);
+            }
+
+            foo();
+
+            Promise.prototype.then = oldThen;

```

Comment:
Is this guaranteed to only run after the completion of `foo()`? I feel like perhaps this should be in a `.then` on `foo()` (or some kind of "after test" hook if we have them, not sure on that) to make sure it is only reset afterwards.

Reply:
I originally tried something like that, but it added a bunch of irrelevant logs to the output because of activity in other tests. Every test in this file is kicked off synchronously and then all of their Promise microtasks queue up to do deferred work, so the output is somewhat interleaved. In this test case, the thing we care about happens synchronously, so we can unset our override immediately. I verified that this test failed without the corresponding C++ change.